### PR TITLE
Implement p2p message

### PR DESCRIPTION
**Pull request by GNUp**

### CODE REVIEW

File: src/p2pMessage.ts [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+}
+
+export class HandshakeMessage {
+    private body: SyncMessage | AckMessage;
+
+    constructor(body: SyncMessage | AckMessage) {
+        this.body = body;
+    }
+
+    toEncodeObject(): Array<any> | number | string {
```
Comment (sgkim126) [inline src/p2pMessage.ts]: When does it return `number` or `string`?

File: src/p2pMessage.ts [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+
+export class AckMessage {
+    private version: number;
+
+    constructor(version: number) {
+        this.version = version;
+    }
+
+    protocolId(): number {
+        return 1;
```
Comment (sgkim126) [inline src/p2pMessage.ts]: return MessageType.ACK_ID;

File: src/p2pMessage.ts [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+        const data = decodedbytes[4];
+
+        if (protocolId === MessageType.ENCRYPTED_ID) {
+            return new ExtensionMessage(
+                version,
+                extensionName,
+                extensionVersion,
+                { type: "encrypted", data }
+            );
+        } else if (protocolId === MessageType.UNENCRYPTED_ID) {
```
Comment (sgkim126) [inline src/p2pMessage.ts]: It is better not to use `else`, if the preceding block always returns.

File: src/p2pMessage.ts [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+    }
+
+    protocolId(): number {
+        switch (this.body.type) {
+            case "request":
+                return MessageType.REQUEST_ID;
+            case "allowed":
+                return MessageType.ALLOWED_ID;
+            case "denied":
+                return MessageType.DENIED_ID;
```
Comment (sgkim126) [inline src/p2pMessage.ts]: Please add `default` case.

File: src/p2pMessage.ts [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+    }
+
+    item_count(): number {
+        switch (this.body.type) {
+            case "request":
+                return COMMON + 2;
+            case "allowed":
+                return COMMON + 1;
+            case "denied":
+                return COMMON;
```
Comment (sgkim126) [inline src/p2pMessage.ts]: Please add `default` case

File: src/p2pMessage.ts [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+                ];
+            case "allowed":
+                return [
+                    this.version,
+                    this.protocolId(),
+                    this.seq,
+                    this.body.version
+                ];
+            case "denied":
+                return [this.version, this.protocolId(), this.seq];
```
Comment (sgkim126) [inline src/p2pMessage.ts]: `default`...

File: src/p2pMessage.ts [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+        this.extensionVersion = extensionVersion;
+        this.data = data;
+    }
+
+    protocolId(): number {
+        switch (this.data.type) {
+            case "encrypted":
+                return MessageType.ENCRYPTED_ID;
+            case "unencrypted":
+                return MessageType.UNENCRYPTED_ID;
```
Comment (sgkim126) [inline src/p2pMessage.ts]: `default`

File: src/p2pMessage.ts (hunk 1 above)
Comment (GNUp) [reply src/p2pMessage.ts]: I removed it.

File: src/p2pMessage.ts (hunk 2 above)
Comment (GNUp) [reply src/p2pMessage.ts]: I changed it

File: src/p2pMessage.ts (hunk 3 above)
Comment (GNUp) [reply src/p2pMessage.ts]: I resolved it.

File: src/p2pMessage.ts (hunk 4 above)
Comment (GNUp) [reply src/p2pMessage.ts]: I added it.

File: src/p2pMessage.ts (hunk 5 above)
Comment (GNUp) [reply src/p2pMessage.ts]: I added it.

File: src/p2pMessage.ts (hunk 7 above)
Comment (GNUp) [reply src/p2pMessage.ts]: I added it.

**Pull request merged**